Linear Regression Using AWS Sagemaker (#115)
ippontech/blog-usa: posts/ml-sagemaker.md

---
authors:
- Mallik Sambaraju
tags:
- Machine Learning
- Data Science
- AWS
- Sagemaker
- Python
date: 2019-03-22T12:24:13.000Z
title: "Linear Regression using AWS Sagemaker"
image: 
---

> Various industries such as Banking, Manufacturing, Insurance, Health, and Defence are using Machine Learning to solve many use cases such as Fraud Detection, Improving Healthcare, Personal security, Providing more secured transportation e.t.c. Recently [Ippon Technologies](https://en.ippon.tech) sent me to a summit on Data Science where I learnt about how organizations are using Machine Learning to improve their business and productivity. During the course of the summit many service and cloud providers presented various tools, libraries, algorithms and platforms for developing Machine Learning applications. One of the services introduced to me was Sagemaker by AWS. In this blog I will talk about How I implemented a basic regression model.

Reviewer: ```suggestion
Various industries such as Banking, Manufacturing, Insurance, Health, and Defense are using Machine Learning to solve many use cases such as Fraud Detection, Improving Healthcare, Personal security, Providing more secured transportation etc. Recently, [Ippon Technologies](https://en.ippon.tech) sent me to a summit on Data Science where I learnt about how organizations are using Machine Learning to improve their business and productivity. During the course of the summit many service and cloud providers presented various tools, libraries, algorithms and platforms for developing Machine Learning applications. One of the services introduced to me was Sagemaker by AWS. In this blog I will talk about How I implemented a basic regression model.
```

# Machine Learning Process

> Typically Machine Learning process is an iterative process, it starts with identifying a use case to solve. Some of the steps involved in this process are as follows. This could change based on the use case you are trying to solve.

Reviewer: ```suggestion
Typically Machine Learning process is an iterative process; it starts with identifying a use case to solve. Some of the steps involved in this process are as follows. This could change based on the use case you are trying to solve.
```

* Identify and Obtain Data.
* Pre-process and Prepare Data.
* Train a Model using the Prepared Data.
* Evaluate the Model.
* Deploy the Model.
* Obtain Feedback of the Model.

Each of the above steps are iterative by themselves. Multiple iterations can happen during each of these steps depending on the data and performance of the model e.t.c.

![Machine Learning Process](https://raw.githubusercontent.com/msambaraju/blog-usa/master/images/2019/03/Machine_Learning.png)

# AWS Sagemaker

> AWS Sagemaker is a fully managed AWS Machine Learning service which helps in building, training and deploying Machine Learning models. It has a rich set API's, built-in algorithms, integration with various popular libraries such as Tensorflow, PyTorch, SparkML e.t.c., developers tools for authoring models, and hosted production environment for deploying the models.

Reviewer: ```suggestion
AWS Sagemaker is a fully managed AWS Machine Learning service which helps in building, training and deploying Machine Learning models. It has a rich set API's, built-in algorithms, integration with various popular libraries such as Tensorflow, PyTorch, SparkML etc. It also features developers tools for authoring models, and hosted production environment for deploying the models.
```

# Regression Model Implementation

> In this example we will develop a regression model to predict body fat percentage based on already labeled data with 14 parameters like Age, Height, Weight, Abdomen circumference e.t.c. This data is widely available on the internet. A more advanced process is to find only the relevent features or parameters to train the model.

Reviewer: ```suggestion
In this example we will develop a regression model to predict body fat percentage based on already labeled data with 14 parameters like age, height, weight, abdomen circumference etc. This data is widely available on the internet. A more advanced process is to find only the relevant features or parameters to train the model.
```
Perhaps show where you got your dataset? Not necessarily required, but if I were to follow your guide, it would be nice to have the dataset handy.

> We will use the Jupyter Notebook authoring environment provided by Sagemaker to Prepare Data, Train and Evaluate Model, Deploy and Test Model. The notebook environment also supports version control systems like CodeCommit or GitHub. Upload any test data used by the Notebooks into the environment. The notebook lets you install any python libraries needed for your model using tools such as pip, run any git commands to push your changes to version control system.

Reviewer: ```suggestion
We will use the Jupyter Notebook authoring environment provided by Sagemaker to Prepare Data, Train and Evaluate a model, and to Deploy and Test the model. The notebook environment also supports version control systems like CodeCommit or GitHub. You can upload any test data used by the Notebooks into the environment. The notebook lets you install any python libraries needed for your model using tools such as pip, run any git commands to push your changes to version control system.
```

> In order to use the Jupyter Notebook, we need to create a Notebook Instance by providing an instance type such as ml.t2.medium. Provide IAM role with proper access control during instance creation.

Reviewer: ```suggestion
In order to use the Jupyter Notebook, we need to create a Notebook Instance by providing an instance type such as `ml.t2.medium`. Provide an IAM role with the proper access control during instance creation.
```

Now create a new Jupyter Notebook to interactively Author, Train, Test and Deploy the Model.

![Jupyter Notebook](https://raw.githubusercontent.com/msambaraju/blog-usa/master/images/2019/03/Jupiter_Notebook_Env.png)

> Prepare the data by reading training dataset from the S3 bucket or from an uploaded file. Format the data into an acceptable format usually in the form of arrays and vectors depending on the algorithm used.

Reviewer: ```suggestion
Prepare the data by reading the training dataset from a S3 bucket or from an uploaded file. Format the data into an acceptable format typically this is in the form of arrays and vectors depending on the algorithm used.
```
Maybe another blog post on the different ML algorithms?

```python
import numpy as np
import boto3
import sagemaker
import io
import sagemaker.amazon.common as smac
import os
import pandas as pd

//Read from csv or someother location like s3.
dataset = pd.read_csv("Bio_Train.csv")

// Determine the features and labels.
feature_dataset = dataset[['Density', 'Age', 'Wt', 'Ht', 'Neck', 'Chest', 
      'ABD', 'Hip', 'Thigh', 'Knee', 'Ankle', 'Biceps', 'Farm', 'Wrist']]
features = np.array(feature_dataset.values).astype('float32')

label_dataset= dataset[['BFat']]
labels = np.array(label_dataset.values).astype('float32')
labels_vec = np.squeeze(np.asarray(labels))

```

Upload the prepared data into S3 bucket.

``` python

buffer = io.BytesIO()
smac.write_numpy_to_dense_tensor(buffer, features, labels_vec)
buffer.seek(0)

key = 'linearregression'
boto3.resource('s3').Bucket(bucket).Object(os.path.join(prefix, 'train', key)).upload_fileobj(buf)
s3_training_data_location = 's3://{}/{}/train/{}'.format(bucket, prefix, key)

```

> Fetch the container with proper algorithm to use from the list of pre-defined Sagemaker algorithms or give your own custom container to support custom algorithms. In this case linear-learner algorithm is used which is a pre-defined algorithm.

Reviewer: ```suggestion
Fetch the container with the corresponding algorithm to use from the list of pre-defined Sagemaker algorithms or give your own custom container to support custom algorithms. In this case linear-learner algorithm is used which is a pre-defined algorithm.
```

 ml.c5.18xlarge, ml.p3.2xlarge, ml.m5.xlarge, ml.m4.10xlarge, ml.m5.12xlarge, ml.m4.xlarge,
ml.m5.24xlarge, ml.m4.2xlarge, ml.p2.8xlarge, ml.m5.2xlarge, ml.p3.8xlarge, ml.m4.4xlarge]
```

``` python
from sagemaker import get_execution_role

role = get_execution_role()

sagemaker_session = sagemaker.Session()

// Provide the container, role, instance type and model output location
linear = sagemaker.estimator.Estimator(linear_container,
                                       role=role, 
                                       train_instance_count=1, 
                                       train_instance_type='ml.c4.xlarge',
                                       output_path=output_location,
                                       sagemaker_session=sagemaker_session)

// Provide the number of features identified during data preparation
// Provide the predictor_type 

linear.set_hyperparameters(feature_dim=14,
                           mini_batch_size=4,
                           predictor_type='regressor')

// Train the model using the previously prepared test data and validate the 
//data by providing the validation data.

linear.fit({'train': s3_training_data_location})

```

> Deploy the Trained Model using the Sagemaker API. Provide instance type and instance count as required. Once the deployment is complete the test data is used to test the deployed application. Once tha Model is deployed a http endpoint is generated which is used by other applications such as a lambda function as part of a streaming application or a synchronous application. To support varying  loads the application can be autoscaled.

Reviewer: ```suggestion
Deploy the Trained Model using the Sagemaker API. Provide the instance type and instance count as required. Once the deployment is complete, the test data is used to test the deployed application. Once the Model is deployed an http endpoint is generated which is used by other applications such as a lambda function as part of a streaming application or a synchronous application. To support varying  loads the application can be autoscaled.
```
where do you find the Sagemaker API?

Did you mean _a synchronous_ or did you mean asynchronous?

  40.6    24.     37.1    30.1    18.2  ]

Prediction: {'predictions': [{'score': 20.166656494140625}]}
```

The deployed application can be invoked from a custom application using the AWS Sagemaker API's as below.

``` java
//Process input data and fetch the input for the model.
String csvLine = "1.061,33.,211.75,73.5,40.,106.2,100.5,109.,65.8,40.6,24.,37.1,30.1,18.2";
AmazonSageMakerRuntime client = AmazonSageMakerRuntimeClientBuilder.standard().
				withRegion(Regions.US_EAST_1).withCredentials
                             (new AWSStaticCredentialsProvider(creds)).build();
		
InvokeEndpointRequest invokeEndpointRequest = new InvokeEndpointRequest();
ByteBuffer buf = ByteBuffer.wrap(csvLine.getBytes());
invokeEndpointRequest.setContentType("text/csv");
invokeEndpointRequest.setAccept("application/json");
invokeEndpointRequest.setBody(buf);
// provide appropriate endpoint name.
invokeEndpointRequest.setEndpointName("ippon-sagemaker-regression-v1");		
InvokeEndpointResult result = client.invokeEndpoint(invokeEndpointRequest);
String responseJson = new String(result.getBody().array());
//Add business logic based on the prediction received.
```

> 

Reviewer: What were the results of the tutorial you did?

# Conclusion

> AWS Sagemaker provides capabilities to Author, Train, Deploy and Monitor Machine Learning applications using wide variety of algorithms, libraries and infrastructure. This can help Data Engineers and Data Scientists to come up with quality models.

Reviewer: Good ending, try to give a call-to-action to include Ippon/have the reader do something after having learned this cool thing.

---
authors:
- Mallik Sambaraju
tags:
- Machine Learning
- Data Science
- AWS
- Sagemaker
- Python
date: 2019-03-22T12:24:13.000Z
title: "Regression using AWS Sagemaker"
image: 
---

> Various industries such as Banking, Manufacturing, Insurance, Health, and Defence are using Machine Learning to solve many use cases such as Fraud Detection, Improving Healthcare, Personal security, Providing more secured transportation etc. Recently [Ippon Technologies](https://en.ippon.tech) sent me to a summit on Data Science where I learnt about how organizations are using Machine Learning to improve their business and productivity. During the course of the summit many service and cloud providers presented various tools, libraries, algorithms and platforms for developing Machine Learning applications. One of the services introduced to me was Sagemaker by AWS. In this blog I will talk about How I implemented a basic regression model.

Reviewer 2: Defence -> Defense
Fraud Detection -> fraud detection
Improving Healthcare -> improving healthcare
Personal -> personal
Providing -> providing
How -> how

# Machine Learning Process
Typically Machine Learning process is an iterative process; it starts with identifying a use case to solve. Some of the steps involved in this process are as follows. This could change based on the use case you are trying to solve.

* Identify and Obtain Data.
* Pre-process and Prepare Data.
* Train a Model using the Prepared Data.
* Evaluate the Model.
* Deploy the Model.
* Obtain Feedback of the Model.

Each of the above steps are iterative by themselves. Multiple iterations can happen during each of these steps depending on the data and performance of the model etc.

![Machine Learning Process](https://raw.githubusercontent.com/msambaraju/blog-usa/master/images/2019/03/Machine_Learning.png)

# AWS Sagemaker

> AWS Sagemaker is a fully managed AWS Machine Learning service which helps in building, training and deploying Machine Learning models. It has a rich set API's, built-in algorithms, integration with various popular libraries such as Tensorflow, PyTorch, SparkML etc. It also developers tools for authoring models, and hosted production environment for deploying the models.

Reviewer 2: built-in algorithms, integration with -> built-in algorithms, and integration with
It also developers -> It is also a developers
and hosted production environment -> and hosting a production environment

# Regression Model Implementation
In this example we will develop a regression model to predict body fat percentage based on already labeled data with 14 parameters like age, height, weight, abdomen circumference etc. This data is widely available ([Dataset Location1](http://lib.stat.cmu.edu/datasets/bodyfat) or [Dataset Location2](http://wiki.stat.ucla.edu/socr/index.php/SOCR_Data_BMI_Regression#References)) on the internet. A more advanced process is to find only the relevant features or parameters to train the model.

> We will use the Jupyter Notebook authoring environment provided by Sagemaker to Prepare Data, Train and Evaluate a model, to Deploy and Test the model. The notebook environment also supports version control systems like CodeCommit or GitHub. You can upload any test data used by the Notebooks into the environment. The notebook lets you install any python libraries needed for your model using tools such as pip, run any git commands to push your changes to version control system.

Reviewer 2: Prepare Data, Train and Evaluate a model, to Deploy and Test -> prepare data, train and evaluate a model, and deploy and test
using tools such as pip, run any git commands -> using tools such as pip.  Run any git commands

In order to use the Jupyter Notebook, we need to create a Notebook Instance by providing an instance type such as `ml.t2.medium`. Provide an IAM role with the proper access control during instance creation.

> Now create a new Jupyter Notebook to interactively Author, Train, Test and Deploy the Model.

Reviewer 2: Author, Train, Test and Deploy the Model -> author, train, test and deploy the model

feature_dataset = dataset[['Density', 'Age', 'Wt', 'Ht', 'Neck', 'Chest', 
      'ABD', 'Hip', 'Thigh', 'Knee', 'Ankle', 'Biceps', 'Farm', 'Wrist']]
features = np.array(feature_dataset.values).astype('float32')

label_dataset= dataset[['BFat']]
labels = np.array(label_dataset.values).astype('float32')
labels_vec = np.squeeze(np.asarray(labels))

```

Upload the prepared data into S3 bucket.

``` python

buffer = io.BytesIO()
smac.write_numpy_to_dense_tensor(buffer, features, labels_vec)
buffer.seek(0)

key = 'linearregression'
boto3.resource('s3').Bucket(bucket).Object(os.path.join(prefix, 'train', key)).upload_fileobj(buf)
s3_training_data_location = 's3://{}/{}/train/{}'.format(bucket, prefix, key)

```

Fetch the container with the corresponding algorithm to use from the list of pre-defined Sagemaker algorithms or give your own custom container to support custom algorithms. In this case pre-defined linear-learner algorithm is used which supports both classification and regression problems.

``` python

from sagemaker.amazon.amazon_estimator import get_image_uri
linear_container = get_image_uri(boto3.Session().region_name, 'linear-learner')

```

> Now train the Model using the container and the training data previously prepared. Create a new instance for training the Model, provide the instance type needed. The trained Model is stored in the S3 bucket as a tar file so provide S3 bucket details.

Reviewer 2: Model -> model

[Sagemaker Java SDK](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/)

``` java
//Process input data and fetch the input for the model.
String csvLine = "1.061,33.,211.75,73.5,40.,106.2,100.5,109.,65.8,40.6,24.,37.1,30.1,18.2";
AmazonSageMakerRuntime client = AmazonSageMakerRuntimeClientBuilder.standard().
				withRegion(Regions.US_EAST_1).withCredentials
                             (new AWSStaticCredentialsProvider(creds)).build();
		
InvokeEndpointRequest invokeEndpointRequest = new InvokeEndpointRequest();
ByteBuffer buf = ByteBuffer.wrap(csvLine.getBytes());
invokeEndpointRequest.setContentType("text/csv");
invokeEndpointRequest.setAccept("application/json");
invokeEndpointRequest.setBody(buf);
// provide appropriate endpoint name.
invokeEndpointRequest.setEndpointName("ippon-sagemaker-regression-v1");		
InvokeEndpointResult result = client.invokeEndpoint(invokeEndpointRequest);
String responseJson = new String(result.getBody().array());
// The respinseJson contains the  prediction as below
// {"predictions": [{"score": 20.166656494140625}]}
// Add business logic based on the prediction received.
```

# Conclusion

> AWS Sagemaker provides capabilities to Author, Train, Deploy and Monitor Machine Learning applications using wide variety of algorithms, libraries and infrastructure. This can help Data Engineers and Data Scientists to come up with quality models.

Reviewer 2: Author, Train, Deploy and Monitor -> author, train, deploy and monitor
